wire/error.go

```diff
@@ -17,19 +187,56 @@ import (
 // differentiate between general io errors such as io.EOF and issues that
 // resulted from malformed messages.
 type MessageError struct {
-	Func        string // Function name
-	Description string // Human readable description of the issue
+	Func        string    // Function name
+	ErrorCode   ErrorCode // Describes the kind of error
+	Description string    // Human readable description of the issue
 }
 
 // Error satisfies the error interface and prints human-readable errors.
-func (e *MessageError) Error() string {
-	if e.Func != "" {
-		return fmt.Sprintf("%v: %v", e.Func, e.Description)
+func (m *MessageError) Error() string {
+	if m.Func != "" {
+		return fmt.Sprintf("%v: [%s] %v", m.Func, m.ErrorCode, m.Description)
+	}
+	return fmt.Sprintf("%v: %v", m.ErrorCode, m.Description)
+}
+
+// messageError creates an Error given a set of arguments.
+func messageError(Func string, c ErrorCode, desc string) *MessageError {
+	return &MessageError{Func: Func, ErrorCode: c, Description: desc}
+}
+
+// Is implements the interface to work with the standard library's errors.Is.
+// If target is a *MessageError, Is returns true if the error codes match.
+// If target is an ErrorCode, Is returns true if the error codes match.
+// Else, Is returns false.
+func (m *MessageError) Is(target error) bool {
+	switch target := target.(type) {
+	case *MessageError:
+		return m.ErrorCode != 0 && m.ErrorCode == target.ErrorCode
+	case ErrorCode:
+		return m.ErrorCode != 0 && target == m.ErrorCode
+	default:
+		return false
+	}
+}
+
+// Unwrap returns the underlying wrapped error if it is not ErrOther.
+// Unwrap returns the ErrorCode. Else, it returns nil.
+func (m *MessageError) Unwrap() error {
+	if m.ErrorCode != 0 {
+		return m.ErrorCode
 	}
-	return e.Description
+	return nil
 }
 
-// messageError creates an error for the given function and description.
-func messageError(f string, desc string) *MessageError {
-	return &MessageError{Func: f, Description: desc}
+// IsBannable returns whether or not the provided error is a bannable error.
+func IsBannable(err error) bool {
+	switch err := err.(type) {
+	case *MessageError:
+		return err.ErrorCode != 0 && err.ErrorCode != ErrUnequalQuantities
```

Comment:
Why is a mismatched witness count something that isn't bannable?  Peers should not be sending improperly serialized transactions.

This is a good example of why this entire function really doesn't belong in wire, in my opinion.